src/aac/genplug.py

```diff
@@ -1,226 +1,275 @@
+
+    Returns:
+        List of TemplateOutputFile objects that contain the compiled templates
+    """
+
+    # ensure model is present and valid, get the plugin name
+    plugin_models = util.get_models_by_type(parsed_models, "model")
+    if len(plugin_models.keys()) != 1:
+        raise GeneratePluginException(
+            "Plugin Arch-as-Code yaml must contain one and only one model."
+        )
+
+    plugin_model = list(plugin_models.values())[0].get("model")
+    plugin_name = plugin_model.get("name")
+    plugin_implementation_name = _convert_to_implementation_name(plugin_name)
+
+    # Ensure that the plugin name has package name prepended to it
+    if not plugin_name.startswith(__package__):
+        plugin_name = "{__package__}-{plugin_name}"
+
+    behaviors = util.search(plugin_model, ["behavior"])
+    commands = _gather_commands(behaviors)
+
+    plugin_templates = load_templates("genplug")
+
+    plugin = {
+        "name": plugin_model.get("name"),
+        "implementation_name": plugin_implementation_name,
+    }
+
+    extensions = [
+        _add_extensions_yaml_string(definition) for definition in _gather_extensions(parsed_models)
+    ]
+
+    template_properties = {"plugin": plugin, "commands": commands, "extensions": extensions}
+    generated_templates = generate_templates(plugin_templates, template_properties)
+
+    # Define which templates we want to overwrite.
+    templates_to_overwrite = ["plugin.py.jinja2", "setup.py.jinja2"]
+
+    # Combine the generated templates into a map of file_name: file_contents
+    files_to_write = []
+    for template_name, template_content in generated_templates.items():
+        file_name = _convert_template_name_to_file_name(template_name, plugin_implementation_name)
+        overwrite = template_name in templates_to_overwrite
+
+        files_to_write.append(TemplateOutputFile(file_name, template_content, overwrite))
+
+    return files_to_write
+
+
+def _write_generated_templates_to_file(
+    generated_files: list[TemplateOutputFile], plug_dir: str
+) -> None:
+    """
+    Write a list of generated files to the target directory.
+
+    Args:
+        generated_files <list>: list of generated files to write to the filesystem
+        plug_dir <str>: the directory to write the generated files to.
+    """
+
+    for generated_file in generated_files:
+        _write_file(
+            plug_dir,
+            generated_file.file_name,
+            generated_file.overwrite,
+            generated_file.template,
+        )
+
+
+def _convert_template_name_to_file_name(template_name: str, plugin_name: str) -> str:
+    """
+    Convert template names to pythonic file names.
+
+    Args:
+        template_name <str>: The template's name
+        plugin_name <str>: The plugin's name
+    Returns:
+        A string containing the personalized/pythonic file name.
+    """
+
+    #  we're using that genplug will only ever generate python templates
```

Comment:
Should we add this as an optional parameter so that, if we want to generate other types of templates (i.e. `yaml` immediately comes to mind as a possible example) we can do so using this same function? Alternatively, could we just get this from the template name without assuming any particular file extension?